File: sdk/js/packages/crypto/src/shielded/keys.ts
// Copyright 2026 TOS Blockchain Teams. SPDX-License-Identifier: LGPL-2.0-or-later

/**
 * Section 2: the V1 key hierarchy.
 *
 * The mnemonic is the root and nothing else is stored. Every key is recomputed
 * from the seed and a note index, which is what makes mnemonic-only recovery
 * possible: a restored wallet does not reconstruct a keystore, it recomputes
 * the keys and asks the chain which ones were used.
 *
 * Everything below the instance master is derived from `poolInstanceMaster`,
 * never from the mnemonic-global `poolMaster`. That is what stops the same
 * mnemonic and index from producing the same key on two pools, and it is why
 * the execution domain -- a deployment fact -- reaches this far into a wallet.
 */

import { sha256 } from "@noble/hashes/sha2";
import { shake256 } from "@noble/hashes/sha3";
import { ml_dsa44 } from "@noble/post-quantum/ml-dsa";
import { ml_kem768 } from "@noble/post-quantum/ml-kem";

import { MODULUS } from "./generated/poseidon2-params.js";
import { type Fr, h7 } from "./poseidon2.js";

/** Frozen sizes, sections 2.2 and 2.3. */
export const MLKEM_ENCAPSULATION_KEY_BYTES = 1184;
export const MLKEM_CIPHERTEXT_BYTES = 1088;
export const MLDSA_PUBLIC_KEY_BYTES = 1312;
export const MLDSA_SIGNATURE_BYTES = 2420;
export const SHARED_SECRET_BYTES = 32;

/** Section 9.4: exactly 28 ASCII bytes, and it never varies. */
export const SIGNATURE_CONTEXT = new TextEncoder().encode("TOS-SHIELDED-POOL-MLDSA44-v1");

const ascii = (text: string): Uint8Array => new TextEncoder().encode(text);

function concat(...parts: Uint8Array[]): Uint8Array {
  const total = parts.reduce((sum, part) => sum + part.length, 0);
  const out = new Uint8Array(total);
  let at = 0;
  for (const part of parts) {
    out.set(part, at);
    at += part.length;
  }
  return out;
}

function shake(parts: Uint8Array[], length: number): Uint8Array {
  return shake256(concat(...parts), { dkLen: length });
}

/** A field element as its 32 canonical big-endian bytes. */
export function frBe32(value: Fr): Uint8Array {
  const out = new Uint8Array(32);
  let rest = ((value % MODULUS) + MODULUS) % MODULUS;
  for (let at = 31; at >= 0; at -= 1) {
    out[at] = Number(rest & 0xffn);
    rest >>= 8n;
  }
  return out;
}

/** The inverse, rejecting anything at or above the modulus. */
export function frFromBe32(bytes: Uint8Array): Fr {
  if (bytes.length !== 32) {
    throw new Error(`a field element is 32 bytes, not ${bytes.length}`);
  }
  let value = 0n;
  for (const byte of bytes) {
    value = (value << 8n) | BigInt(byte);
  }
  if (value >= MODULUS) {
    throw new Error("a field element at or above the modulus");
  }
  return value;
}

/** `reduce_to_field(SHA256(tag || body))`, section 3 and section 8's shape. */
export function taggedHash(tag: string, body: Uint8Array): Fr {
  const digest = sha256(concat(ascii(tag), body));
  let value = 0n;
  for (const byte of digest) {
    value = (value << 8n) | BigInt(byte);
  }
  return value % MODULUS;
}

/** Public inputs 6, 7 and 8. */
export const outputDataHash = (bytes: Uint8Array): Fr =>
  taggedHash("TOS-SHIELDED-OUTPUT-DATA-v1", bytes);

/** Section 4.1, and public inputs 9 and 10. */
export const pqAuthKeyHash = (publicKey: Uint8Array): Fr =>
  taggedHash("TOS-SHIELDED-MLDSA44-PK-v1", publicKey);

/** Public input 13. A transfer's is zero and never reaches this. */
export const publicRecipientHash = (account: Uint8Array): Fr =>
  taggedHash("TOS-SHIELDED-RECIPIENT-v1", account);

/**
 * Section 8, and public input 15. A deployment fact: it binds the chain and
 * the particular pool account, so a proof made for one pool cannot be replayed
 * against another.
 */
export function executionDomain(globalId: number, poolAccount: Uint8Array): Fr {
  const body = new Uint8Array(4 + 1 + 32 + 2);
  new DataView(body.buffer).setInt32(0, globalId, false);
  body[4] = 0;
  body.set(poolAccount, 5);
  body[37] = 0;
  body[38] = 1;
  return taggedHash("TOS-SHIELDED-EXEC-v1", body);
}

/** Section 15.1, and public input 16. Poseidon2, not SHA-256: its operands are already field elements. */
export const recoveryTemplateHash = (ownerCommitment: Fr, recoveryDataHash: Fr): Fr =>
  h7("RECOVERY-TEMPLATE", [ownerCommitment, recoveryDataHash, 0n, 0n, 0n, 0n, 0n]);

/** The mnemonic-global base, never used to derive a key directly. */
export const poolMaster = (tosSeed: Uint8Array): Uint8Array =>
  shake([ascii("TOS-SHIELDED-POOL-MASTER-v1"), tosSeed], 64);

function beU64(value: bigint): Uint8Array {
  const out = new Uint8Array(8);
  new DataView(out.buffer).setBigUint64(0, value, false);
  return out;
}

/** Everything a wallet holds for one pool on one chain. */
export class PoolInstance {
  private readonly master: Uint8Array;

  readonly executionDomain: Fr;

  constructor(tosSeed: Uint8Array, executionDomain: Fr) {
    this.master = shake(
      [ascii("TOS-SHIELDED-POOL-INSTANCE-v1"), poolMaster(tosSeed), frBe32(executionDomain)],
      64,
    );
    this.executionDomain = executionDomain;
  }

  /**
   * Section 2.1. A derivation that reduces to zero is not usable, so attempt
   * zero is the bare bytes and each retry appends its own counter. Returning
   * the first non-zero result keeps the mapping from index to key
   * deterministic, which a recovering wallet depends on.
   */
  ownerNfKey(noteKeyIndex: bigint): Fr {
    const index = beU64(noteKeyIndex);
    for (let retry = 0; retry <= 255; retry += 1) {
      const parts = [ascii("TOS-SHIELDED-OWNER-NF-v1"), this.master, index];
      if (retry > 0) {
        parts.push(Uint8Array.of(retry));
      }
      const wide = shake(parts, 64);
      let value = 0n;
      for (const byte of wide) {
        value = (value << 8n) | BigInt(byte);
      }
      const reduced = value % MODULUS;
      if (reduced !== 0n) {
        return reduced;
      }
    }
    throw new Error(`no non-zero owner-nullifier key for index ${noteKeyIndex}`);
  }

  ownerNfKeyHash(noteKeyIndex: bigint): Fr {
    return h7("OWNER-NF-HASH", [this.ownerNfKey(noteKeyIndex), 0n, 0n, 0n, 0n, 0n, 0n]);
  }

  /** Section 2.2: the 64-byte (d, z) seed FIPS 203 permits storing. */
  mlkemSeed(): { d: Uint8Array; z: Uint8Array } {
    return {
      d: shake([ascii("TOS-SHIELDED-MLKEM-D-v1"), this.master], 32),
      z: shake([ascii("TOS-SHIELDED-MLKEM-Z-v1"), this.master], 32),
    };
  }

  mlkemKeys(): { publicKey: Uint8Array; secretKey: Uint8Array } {
    const { d, z } = this.mlkemSeed();
    return ml_kem768.keygen(concat(d, z));
  }

  mlkemEncapsulationKey(): Uint8Array {
    return this.mlkemKeys().publicKey;
  }

  /** Section 2.3: a different ML-DSA key per note, from the same index. */
  mldsaSeed(noteKeyIndex: bigint): Uint8Array {
    return shake([ascii("TOS-SHIELDED-MLDSA44-KEY-v1"), this.master, beU64(noteKeyIndex)], 32);
  }

  mldsaKeys(noteKeyIndex: bigint): { publicKey: Uint8Array; secretKey: Uint8Array } {
    return ml_dsa44.keygen(this.mldsaSeed(noteKeyIndex));
  }

  mldsaPublicKey(noteKeyIndex: bigint): Uint8Array {
    return this.mldsaKeys(noteKeyIndex).publicKey;
  }

  pqAuthKeyHash(noteKeyIndex: bigint): Fr {
    return pqAuthKeyHash(this.mldsaPublicKey(noteKeyIndex));
  }

  /** Section 9.4: over the digest's 32 big-endian bytes, under the context. */
  signIntent(noteKeyIndex: bigint, intentDigest: Fr): Uint8Array {
    const { secretKey } = this.mldsaKeys(noteKeyIndex);
    return ml_dsa44.sign(secretKey, frBe32(intentDigest), SIGNATURE_CONTEXT);
  }

  /** Only the holder of the mnemonic can do this. */
  decapsulate(ciphertext: Uint8Array): Uint8Array {
    return ml_kem768.decapsulate(ciphertext, this.mlkemKeys().secretKey);
  }
}

/** Section 3 step 1: encapsulate to a recipient's key. */
export function encapsulateTo(encapsulationKey: Uint8Array): {
  sharedSecret: Uint8Array;
  cipherText: Uint8Array;
} {
  return ml_kem768.encapsulate(encapsulationKey);
}

File: tools/shielded-pool-wallet/src/keys.rs
/*
 * Copyright (C) 2026-2026 TOS Blockchain Teams.
 *
 * Licensed under the GNU General Public License v3.0.
 * See the LICENSE file in the root of this repository.
 */
//! Section 2: the V1 key hierarchy.
//!
//! The mnemonic is the root and nothing else is stored. Every key a wallet
//! ever needs is recomputed from the seed and a `note_key_index`, which is
//! what makes mnemonic-only recovery possible: a restored wallet does not
//! reconstruct a keystore, it recomputes the keys and asks the chain which
//! ones were used.
//!
//! Everything below the instance master is derived from
//! `pool_instance_master`, never from the mnemonic-global `pool_master`. That
//! is what stops the same mnemonic and index from producing the same key on
//! two pools or on mainnet and testnet, and it is why the execution domain --
//! a deployment fact -- reaches this far into the wallet.

use ark_ff::{BigInteger, PrimeField};
use fips203::ml_kem_768;
use fips203::traits::{Decaps as _, Encaps as _, KeyGen as _, SerDes as _};
use fips204::ml_dsa_44;
use fips204::traits::{KeyGen as _, SerDes as _, Signer as _};
use sha3::digest::{ExtendableOutput, Update, XofReader};
use sha3::Shake256;
use shielded_pool_circuit::domains::h7;
use shielded_pool_circuit::field::Fr;

use crate::error::{Error, Result};

/// Frozen sizes, section 2.2 and 2.3.
pub const MLKEM_ENCAPSULATION_KEY_BYTES: usize = 1184;
pub const MLKEM_CIPHERTEXT_BYTES: usize = 1088;
pub const MLDSA_PUBLIC_KEY_BYTES: usize = 1312;
pub const MLDSA_SIGNATURE_BYTES: usize = 2420;
pub const SHARED_SECRET_BYTES: usize = 32;

/// `SHAKE256(input, n)`.
fn shake(parts: &[&[u8]], out: &mut [u8]) {
    let mut hasher = Shake256::default();
    for part in parts {
        hasher.update(part);
    }
    hasher.finalize_xof().read(out);
}

/// A field element as its 32 canonical big-endian bytes.
pub fn fr_be32(value: Fr) -> [u8; 32] {
    let digits = value.into_bigint().to_bytes_be();
    let mut out = [0u8; 32];
    out[32 - digits.len()..].copy_from_slice(&digits);
    out
}

/// The mnemonic-global base. Derived once per seed, and never used to derive
/// a key directly.
pub fn pool_master(tos_seed: &[u8; 32]) -> [u8; 64] {
    let mut out = [0u8; 64];
    shake(&[b"TOS-SHIELDED-POOL-MASTER-v1", tos_seed], &mut out);
    out
}

/// Everything a wallet holds for one pool on one chain.
///
/// Two execution domains from the same mnemonic share no key material below
/// this point, which is section 19 gate 30 and the reason this type exists
/// rather than a bag of free functions taking a domain.
pub struct PoolInstance {
    master: [u8; 64],
    domain: Fr,
}

impl PoolInstance {
    pub fn new(tos_seed: &[u8; 32], execution_domain: Fr) -> Self {
        let mut master = [0u8; 64];
        shake(
            &[b"TOS-SHIELDED-POOL-INSTANCE-v1", &pool_master(tos_seed), &fr_be32(execution_domain)],
            &mut master,
        );
        Self { master, domain: execution_domain }
    }

    pub fn execution_domain(&self) -> Fr {
        self.domain
    }

    /// Section 2.1. The retry rule is the interesting part: a derivation that
    /// reduces to zero is not usable, so attempt zero is the bare bytes and
    /// each retry appends its own counter. Returning the first non-zero result
    /// keeps the mapping from index to key deterministic, which is what a
    /// recovering wallet depends on.
    pub fn owner_nf_key(&self, note_key_index: u64) -> Result<Fr> {
        let index = note_key_index.to_be_bytes();
        for retry in 0u16..=255 {
            let mut wide = [0u8; 64];
            if retry == 0 {
                shake(&[b"TOS-SHIELDED-OWNER-NF-v1", &self.master, &index], &mut wide);
            } else {
                let counter = [retry as u8];
                shake(&[b"TOS-SHIELDED-OWNER-NF-v1", &self.master, &index, &counter], &mut wide);
            }
            let value = Fr::from_be_bytes_mod_order(&wide);
            if value != Fr::from(0u64) {
                return Ok(value);
            }
        }
        Err(Error::KeyDerivation(format!(
            "no non-zero owner-nullifier key for index {note_key_index} in 256 attempts"
        )))
    }

    /// Section 2.1, and what a note commitment binds rather than the key.
    pub fn owner_nf_key_hash(&self, note_key_index: u64) -> Result<Fr> {
        let key = self.owner_nf_key(note_key_index)?;
        Ok(h7(
            "OWNER-NF-HASH",
            &[
                key,
                Fr::from(0u64),
                Fr::from(0u64),
                Fr::from(0u64),
                Fr::from(0u64),
                Fr::from(0u64),
                Fr::from(0u64),
            ],
        ))
    }

    /// Section 2.2. One long-lived ML-KEM key per execution domain, stored as
    /// the 64-byte `(d, z)` seed FIPS 203 permits, so a wallet still holds
    /// only a mnemonic. Long-lived within a domain is what lets a restored
    /// wallet decrypt everything ever sent to it.
    pub fn mlkem_seed(&self) -> ([u8; 32], [u8; 32]) {
        let mut d = [0u8; 32];
        let mut z = [0u8; 32];
        shake(&[b"TOS-SHIELDED-MLKEM-D-v1", &self.master], &mut d);
        shake(&[b"TOS-SHIELDED-MLKEM-Z-v1", &self.master], &mut z);
        (d, z)
    }

    pub fn mlkem_keys(&self) -> (ml_kem_768::EncapsKey, ml_kem_768::DecapsKey) {
        let (d, z) = self.mlkem_seed();
        ml_kem_768::KG::keygen_from_seed(d, z)
    }

    pub fn mlkem_encapsulation_key(&self) -> Result<[u8; MLKEM_ENCAPSULATION_KEY_BYTES]> {
        Ok(self.mlkem_keys().0.into_bytes())
    }

    /// Section 2.3. A different ML-DSA key per note, from the same index as
    /// the owner-nullifier key, so a descriptor carries no stable owner hash.
    pub fn mldsa_seed(&self, note_key_index: u64) -> [u8; 32] {
        let mut xi = [0u8; 32];
        shake(
            &[b"TOS-SHIELDED-MLDSA44-KEY-v1", &self.master, &note_key_index.to_be_bytes()],
            &mut xi,
        );
        xi
    }

    pub fn mldsa_keys(&self, note_key_index: u64) -> (ml_dsa_44::PublicKey, ml_dsa_44::PrivateKey) {
        ml_dsa_44::KG::keygen_from_seed(&self.mldsa_seed(note_key_index))
    }

    pub fn mldsa_public_key(&self, note_key_index: u64) -> Result<[u8; MLDSA_PUBLIC_KEY_BYTES]> {
        Ok(self.mldsa_keys(note_key_index).0.into_bytes())
    }

    /// Section 4.1, and public inputs 9 and 10.
    pub fn pq_auth_key_hash(&self, note_key_index: u64) -> Result<Fr> {
        Ok(shielded_pool_circuit::wire::pq_auth_key_hash(&self.mldsa_public_key(note_key_index)?))
    }

    /// Section 9.4: the signature a spend carries, over the intent digest's
    /// 32 big-endian bytes under the fixed 28-byte context.
    pub fn sign_intent(
        &self,
        note_key_index: u64,
        intent_digest: Fr,
    ) -> Result<[u8; MLDSA_SIGNATURE_BYTES]> {
        let (_, secret) = self.mldsa_keys(note_key_index);
        secret
            .try_sign(&fr_be32(intent_digest), crate::SIGNATURE_CONTEXT)
            .map_err(|error| Error::KeyDerivation(format!("ML-DSA-44 signing: {error}")))
    }

    /// Section 3 step 1: encapsulate to a recipient's key.
    pub fn encapsulate_to(
        encapsulation_key: &[u8; MLKEM_ENCAPSULATION_KEY_BYTES],
    ) -> Result<([u8; SHARED_SECRET_BYTES], [u8; MLKEM_CIPHERTEXT_BYTES])> {
        let key = ml_kem_768::EncapsKey::try_from_bytes(*encapsulation_key)
            .map_err(|error| Error::KeyDerivation(format!("ML-KEM-768 public key: {error}")))?;
        let (secret, ciphertext) = key
            .try_encaps()
            .map_err(|error| Error::KeyDerivation(format!("ML-KEM-768 encapsulation: {error}")))?;
        Ok((secret.into_bytes(), ciphertext.into_bytes()))
    }

    /// The other half, which only the holder of the mnemonic can do.
    pub fn decapsulate(
        &self,
        ciphertext: &[u8; MLKEM_CIPHERTEXT_BYTES],
    ) -> Result<[u8; SHARED_SECRET_BYTES]> {
        let (_, decaps) = self.mlkem_keys();
        let ciphertext = ml_kem_768::CipherText::try_from_bytes(*ciphertext)
            .map_err(|error| Error::KeyDerivation(format!("ML-KEM-768 ciphertext: {error}")))?;
        let secret = decaps
            .try_decaps(&ciphertext)
            .map_err(|error| Error::KeyDerivation(format!("ML-KEM-768 decapsulation: {error}")))?;
        Ok(secret.into_bytes())
    }
}
